---
title: "DADA2 Package Tutorial"
author: "Hao"
date: "7/8/2019"
output:   
  html_document:
    number_sections: FALSE
    toc: true
    fig_width: 8
    fig_height: 6
    theme: paper
    highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```


# **Introduction**

**Accurate, high-resolution sample inference from amplicon sequencing data.**

Authors: Benjamin Callahan <benjamin.j.callahan at gmail.com>, Paul McMurdie, Susan Holmes

[Main Website](https://benjjneb.github.io/dada2/index.html)

The dada2 package infers exact amplicon sequence variants (ASVs) from high-throughput amplicon sequencing data, replacing the coarser and less accurate OTU clustering approach. The dada2 pipeline takes as input demultiplexed fastq files, and outputs the sequence variants and their sample-wise abundances after removing substitution and chimera errors. Taxonomic classification is available via a native implementation of the RDP naive Bayesian classifier, and species-level assignment to 16S rRNA gene fragments by exact matching. https://bioconductor.org/packages/release/bioc/html/dada2.html


<br><br>

# **DADA2 1.12 RELEASE NOTES**

## **NEW FEATURES**

  - The new `seqComplexity` function quantifies the complexity of sequences in terms of the Shannon richness of their kmers. `plotComplexity` interrogates the distribution of sequences complexities in fastq files, and the `rm.lowcomplex` argument in the `filterAndTrim` function allows filtering of low complexity sequences.

  - The new `removePrimers` function removes forward and reverse primers from sequencing reads, and can orient reads based on the location of the forward primer. Currently we recommend `removePrimers` for use with PacBio CCS data, but external solutions remain recommended for Illumina data.

## **SIGNIFICANT USER-VISIBLE CHANGES**

  - The `dada` function can now accept fastq filenames rather than requiring files be dereplicated and stored into memory first. This allows memory requirements to remain flat when processing large numbers of samples.

  - [Pseudo-pooling, an algorithmic approximation to sample inference from pooled samples](https://benjjneb.github.io/dada2/pseudo.html#pseudo-pooling), now has memory requirements that remain flat with sample number when invoked using filenames, e.g. `dada(fastqFiles, err=err, pool="pseudo")`. We now recommend pseudo-pooling for those interested in detecting singleton ASVs in their samples.

## **BUG FIXES**

  - Pooled sample inference with `dada(..., pool=TRUE)` no longer fails to output the most abundant ASV in the first sample under certain conditions.

  - The `data.frame` returned by `mergePairs` is now properly formatted even when only one sample was provided as a list.


<br><br>

# **Installation**

Binaries for the current release version of DADA2 (1.12) are available from Bioconductor. Note that you must have R 3.6.0 or newer, and [Bioconductor version 3.9](https://www.bioconductor.org/install/), to install the current release from Bioconductor.

```{r, eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("dada2", version = "3.9")
```

If you wish to install the latest and greatest development version, or to install to earlier versions of R, see our [from-source installation instructions](https://benjjneb.github.io/dada2/dada-installation.html).

```{r, eval=F}
install.packages("devtools")
library("devtools")
devtools::install_github("benjjneb/dada2", ref="v1.12") # change the ref argument to get other versions
```

# **DADA2 Pipeline Tutorial (1.12)**

Here we walk through version 1.12 of the DADA2 pipeline on a small multi-sample dataset. Our starting point is a set of Illumina-sequenced paired-end fastq files that have been split (or “demultiplexed”) by sample and from which the barcodes/adapters have already been removed. The end product is an **amplicon sequence variant (ASV) table**, a higher-resolution analogue of the traditional OTU table, which records the number of times each [exact amplicon sequence variant](https://www.nature.com/articles/ismej2017119) was observed in each sample. We also assign taxonomy to the output sequences, and demonstrate how the data can be imported into the popular [phyloseq](https://joey711.github.io/phyloseq/) R package for the analysis of microbiome data.

## **Starting point**

This workflow assumes that your sequencing data meets certain criteria:

  - Samples have been demultiplexed, i.e. split into individual per-sample fastq files.
  - Non-biological nucleotides have been removed, e.g. primers, adapters, linkers, etc.
  - If paired-end sequencing data, the forward and reverse fastq files contain reads in matched order.
  
  
If these criteria are not true for your data (are you sure there aren’t any primers hanging around?) you need to remedy those issues before beginning this workflow. See [the FAQ](https://benjjneb.github.io/dada2/faq.html) for recommendations for some common issues.


## **Getting ready**

Load `dada2` package.

```{r}
library(dada2); packageVersion("dada2")
```

The data we will work with are the same as those used in the [mothur MiSeq SOP](https://www.mothur.org/wiki/MiSeq_SOP). To follow along, download the [example data](http://www.mothur.org/w/images/d/d6/MiSeqSOPData.zip) and unzip. These fastq files were generated by 2x250 Illumina Miseq amplicon sequencing of the V4 region of the 16S rRNA gene from gut samples collected longitudinally from a mouse post-weaning. For now just consider them paired-end fastq files to be processed. Define the following path variable so that it points to the extracted directory on your machine:

```{r}
path <- "MiSeq_SOP" # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path)
```

If the package successfully loaded and your listed files match those here, you are ready to go through the DADA2 pipeline.

Now we read in the names of the fastq files, and perform some string manipulation to get matched lists of the forward and reverse fastq files.

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```


## **Inspect read quality profiles**
We start by visualizing the quality profiles of the forward reads:
```{r}
plotQualityProfile(fnFs[1:2])
```

In gray-scale is a heat map of the frequency of each quality score at each base position. The median quality score at each position is shown by the green line, and the quartiles of the quality score distribution by the orange lines. The red line shows the scaled proportion of reads that extend to at least that position (this is more useful for other sequencing technologies, as Illumina reads are typically all the same lenghth, hence the flat red line).

The forward reads are good quality. We generally advise trimming the last few nucleotides to avoid less well-controlled errors that can arise there. These quality profiles do not suggest that any additional trimming is needed. We will truncate the forward reads at position 240 (trimming the last 10 nucleotides).

Now we visualize the quality profile of the reverse reads:
```{r}
plotQualityProfile(fnRs[1:2])
```

The reverse reads are of significantly worse quality, especially at the end, which is common in Illumina sequencing. This isn’t too worrisome, as DADA2 incorporates quality information into its error model which makes the algorithm robust to lower quality sequence, but trimming as the average qualities crash will improve the algorithm’s sensitivity to rare sequence variants. Based on these profiles, we will truncate the reverse reads at position 160 where the quality distribution crashes.

## **Filter and trim**

Assign the filenames for the filtered fastq.gz files.
```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

We’ll use standard filtering parameters: maxN=0 (DADA2 requires no Ns), truncQ=2, rm.phix=TRUE and maxEE=2. The maxEE parameter sets the maximum number of “expected errors” allowed in a read, which is a better filter than simply averaging quality scores.
```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

## **Learn the Error Rates**

The DADA2 algorithm makes use of a parametric error model (`err`) and every amplicon dataset has a different set of error rates. The `learnErrors` method learns this error model from the data, by alternating estimation of the error rates and inference of sample composition until they converge on a jointly consistent solution. As in many machine-learning problems, the algorithm must begin with an initial guess, for which the maximum possible error rates in this data are used (the error rates if only the most abundant sequence is correct and all the rest are errors).

*The following runs in about 3 minutes on a 2013 Macbook Pro:*
```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
```

It is always worthwhile, as a sanity check if nothing else, to visualize the estimated error rates:
```{r}
plotErrors(errF, nominalQ=TRUE)
```

The error rates for each possible transition (A→C, A→G, …) are shown. Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score. Here the estimated error rates (black line) are a good fit to the observed rates (points), and the error rates drop with increased quality as expected. Everything looks reasonable and we proceed with confidence.

# **Sample Inference**

We are now ready to apply [the core sample inference algorithm](https://www.nature.com/articles/nmeth.3869#methods) to the dereplicated data.
```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
```

Inspecting the returned `dada-class` object:
```{r}
dadaFs[[1]]
```

The DADA2 algorithm inferred 128 true sequence variants from the 1979 unique sequences in the first sample. There is much more to the  dada-class return object than this (see help("dada-class") for some info), including multiple diagnostics about the quality of each denoised sequence variant, but that is beyond the scope of an introductory tutorial.


## **Merge paired reads**

We now merge the forward and reverse reads together to obtain the full denoised sequences. Merging is performed by aligning the denoised forward reads with the reverse-complement of the corresponding denoised reverse reads, and then constructing the merged “contig” sequences. By default, merged sequences are only output if the forward and reverse reads overlap by at least 12 bases, and are identical to each other in the overlap region (but these conditions can be changed via function arguments).
```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

The `mergers` object is a list of `data.frames` from each sample. Each `data.frame` contains the merged `$sequence`, its `$abundance`, and the indices of the `$forward` and `$reverse` sequence variants that were merged. Paired reads that did not exactly overlap were removed by  mergePairs, further reducing spurious output.

## **Construct sequence table**

We can now construct an amplicon sequence variant table (ASV) table, a higher-resolution version of the OTU table produced by traditional methods.
```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

The sequence table is a `matrix` with rows corresponding to (and named by) the samples, and columns corresponding to (and named by) the sequence variants. This table contains 293 ASVs, and the lengths of our merged sequences all fall within the expected range for this V4 amplicon.

## **Remove chimeras**

The core `dada` method corrects substitution and indel errors, but chimeras remain. Fortunately, the accuracy of sequence variants after denoising makes identifying chimeric ASVs simpler than when dealing with fuzzy OTUs. Chimeric sequences are identified if they can be exactly reconstructed by combining a left-segment and a right-segment from two more abundant “parent” sequences.
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)
```

The frequency of chimeric sequences varies substantially from dataset to dataset, and depends on on factors including experimental procedures and sample complexity. Here chimeras make up about 21% of the merged sequence variants, but when we account for the abundances of those variants we see they account for only about 4% of the merged sequence reads.


## **Track reads through the pipeline**

As a final check of our progress, we’ll look at the number of reads that made it through each step in the pipeline:
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

Looks good! We kept the majority of our raw reads, and there is no over-large drop associated with any single step.

