---
title: "DADA2 Package Tutorial"
author: "Hao"
date: "7/8/2019"
output:   
  html_document:
    number_sections: FALSE
    toc: true
    fig_width: 8
    fig_height: 6
    theme: paper
    highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```

# **Introduction**

**Accurate, high-resolution sample inference from amplicon sequencing data.**

Authors: Benjamin Callahan <benjamin.j.callahan at gmail.com>, Paul McMurdie, Susan Holmes

[Main Website](https://benjjneb.github.io/dada2/index.html)

The dada2 package infers exact amplicon sequence variants (ASVs) from high-throughput amplicon sequencing data, replacing the coarser and less accurate OTU clustering approach. The dada2 pipeline takes as input demultiplexed fastq files, and outputs the sequence variants and their sample-wise abundances after removing substitution and chimera errors. Taxonomic classification is available via a native implementation of the RDP naive Bayesian classifier, and species-level assignment to 16S rRNA gene fragments by exact matching. https://bioconductor.org/packages/release/bioc/html/dada2.html


# **DADA2 1.12 RELEASE NOTES**

## **NEW FEATURES**

  - The new `seqComplexity` function quantifies the complexity of sequences in terms of the Shannon richness of their kmers. `plotComplexity` interrogates the distribution of sequences complexities in fastq files, and the `rm.lowcomplex` argument in the `filterAndTrim` function allows filtering of low complexity sequences.

  - The new `removePrimers` function removes forward and reverse primers from sequencing reads, and can orient reads based on the location of the forward primer. Currently we recommend `removePrimers` for use with PacBio CCS data, but external solutions remain recommended for Illumina data.

## **SIGNIFICANT USER-VISIBLE CHANGES**

  - The `dada` function can now accept fastq filenames rather than requiring files be dereplicated and stored into memory first. This allows memory requirements to remain flat when processing large numbers of samples.

  - [Pseudo-pooling, an algorithmic approximation to sample inference from pooled samples](https://benjjneb.github.io/dada2/pseudo.html#pseudo-pooling), now has memory requirements that remain flat with sample number when invoked using filenames, e.g. `dada(fastqFiles, err=err, pool="pseudo")`. We now recommend pseudo-pooling for those interested in detecting singleton ASVs in their samples.

## **BUG FIXES**

  - Pooled sample inference with `dada(..., pool=TRUE)` no longer fails to output the most abundant ASV in the first sample under certain conditions.

  - The `data.frame` returned by `mergePairs` is now properly formatted even when only one sample was provided as a list.






